tsl/src/continuous_aggs/create.c

```diff
@@ -1147,9 +1147,19 @@ get_bucket_width(CAggTimebucketInfo bucket_info)
 				bucket_info.interval->day = bucket_info.interval->month * DAYS_PER_MONTH;
 				bucket_info.interval->month = 0;
 			}
+			int64 MSECS_PER_SEC = 1000;
+			Datum float_usecs_per_sec = DirectFunctionCall1(i8tod, Int64GetDatum(USECS_PER_SEC));
+			Datum float_msecs_per_sec = DirectFunctionCall1(i8tod, Int64GetDatum(MSECS_PER_SEC));
 			Datum epoch = DirectFunctionCall2(interval_part,
 											  PointerGetDatum(cstring_to_text("epoch")),
 											  IntervalPGetDatum(bucket_info.interval));
+			/* Preserve precision if possible for units less than seconds by converting to usecs or
+			 * msecs */
+			if (PG_INT64_MAX / USECS_PER_SEC + PG_INT64_MAX % USECS_PER_SEC > DatumGetFloat8(epoch))
+				epoch = DirectFunctionCall2(float8mul, epoch, float_usecs_per_sec);
+			else if (PG_INT64_MAX / MSECS_PER_SEC + PG_INT64_MAX % MSECS_PER_SEC >
+					 DatumGetFloat8(epoch))
+				epoch = DirectFunctionCall2(float8mul, epoch, float_msecs_per_sec);
```

Comment:
I am a little confused by this: epoch is a float8, and depending on its value we convert it either to milliseconds or microseconds. How does the subsequent code know in which units it is?

Reply:
Thanks a lot for reviewing! You're right to be confused, as we don't know which units it is. Sorry about that. My reasoning was, first try to convert to usec. It's the smallest unit of time supported in intervals so there should be no loss here. If the value is too large to convert to usec, then try to convert to milliseconds, but of course that might not always work.

Reply:
Fixed it by avoid using epoch and converting Interval to integer.